### ISSUE TITLE
Integrate ARAX with Genetics Provider

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#1064]
Request from Marcin von Grotthuss:

> I'd like to ask you if you would be interested in linking our Genetics KP with your ARAX for the drug repurposing use-case. I am asking since we have just updated our KP with gene-phenotype association for 143 traits and I think genetics data might be very informative, particular, for this use-case. I will be happy to show you how our KP works and here is just an example of querying for phenotype to gene associations:
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"EFO:0001360","id":"n00","type":"disease"},{"id":"n01","type":"gene"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query

some example nodes/edges returned from this KP:
```
{
  "knowledge_graph": {
    "edges": [
      {
        "id": "MAGMA_GENE_2184288",
        "score": 5.9463e-157,
        "score_direction": "smaller_is_better",
        "score_name": "MAGMA-pvalue",
        "source_id": "EFO:0001360",
        "target_id": "NCBIGene:6934",
        "type": "associated"
      },
      {
        "id": "INTEGRATED_GENETICS_80254",
        "score": 0.96406529962174,
        "score_direction": "higher_is_better",
        "score_name": "Genetics-quantile",
        "source_id": "EFO:0001360",
        "target_id": "NCBIGene:85236",
        "type": "associated"
      },
      ...
     "nodes": [
      {
        "id": "EFO:0001360",
        "type": "disease"
      },
      {
        "id": "NCBIGene:6934",
        "type": "gene"
      },
      {
        "id": "NCBIGene:3784",
        "type": "gene"
      },
      ...
```

**Comment by amykglen** [comment]
some meta-data for their KG is here: https://translator.broadinstitute.org/genetics_data_provider/predicates

summary:

- curie prefixes: NCBIGene, EFO, GO
- node types: `gene`, `phenotype`/`disease` (either of these will produce the same answer), `pathway`
- predicates: `associated`
- scores provided:
  - MAGMA-pvalue (for gene--phenotype and pathway--phenotype edges)
  - Genetics-quantile (for gene--phenotype edges only)

**Comment by amykglen** [comment]
some more details from Marcin about the edges they provide:

> saying that we are sending back only associations that pass significant level cutoffs, i.e.:
> phenotype-to-gene, MAGMA p-value < 2.5e-6
> phenotype-to-gene, Integrated score > 0.95
> pathway-to-phenotype, MAGMA p-value < 2e-6
> gene-to-phenotype, MAGMA p-value < 0.05
> gene-to-phenotype, Integrated score > 0.80
> phenotype-to-pathway, MAGMA p-value < 0.05

**Comment by saramsey** [comment]
Genetics KP tool home page is: http://www.kp4cd.org/

**Comment by saramsey** [comment]
OUTSTANDING work on this issue, @amykglen.

**Comment by amykglen** [comment]
example API call for their current method of batch querying (equivalent to sending three queries for NCBIGene:5468, NCBIGene:6934, and NCBIGene:162515 separately and combining their results):
```
{
  "message": {
    "query_graph": {
      "edges": [
        {
          "id": "e00",
          "source_id": "n00",
          "target_id": "n03",
          "type": "associated"
        },
        {
          "id": "e00",
          "source_id": "n01",
          "target_id": "n03",
          "type": "associated"
        },
                {
          "id": "e00",
          "source_id": "n02",
          "target_id": "n03",
          "type": "associated"
        }
      ],
      "nodes": [
        {
          "curie": "NCBIGene:5468",
          "id": "n00",
          "type": "gene"
        },
        {
          "curie": "NCBIGene:6934",
          "id": "n01",
          "type": "gene"
        },
        {
          "curie": "NCBIGene:162515",
          "id": "n02",
          "type": "gene"
        },
        {
          "id": "n03",
          "type": "phenotypic_feature"
        }
      ]
    }
  }
}
```

**Comment by amykglen** [comment]
an update from Marcin about the GeneticsKP API:

> I changed our API in the way that now our traits (EFOs) are split into 2 distinct categories, namely "disease" and "phenotypic_features". Also, I removed from our pathway all the GOs that are not marked as biological_process in the gene ontology. We will add back those removed GOs when we switch to TRAPI 1.0. And just for reference, here are the examples of 8 different queries supported by our API:
> 
> find associated diseases for a PPRG gene (NCBIGene:5468)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"NCBIGene:5468","id":"n00","type":"gene"},{"id":"n01","type":"disease"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated quantitative traits for a PPRG gene (NCBIGene:5468)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"NCBIGene:5468","id":"n00","type":"gene"},{"id":"n01","type":"phenotypic_feature"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated diseases for pathway (GO:0061088)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"GO:0061088","id":"n00","type":"pathway"},{"id":"n01","type":"disease"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated quantitative traits for pathway (GO:0061088)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"GO:0061088","id":"n00","type":"pathway"},{"id":"n01","type":"phenotypic_feature"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated genes for type 2 diabetes (EFO:0001360)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"EFO:0001360","id":"n00","type":"disease"},{"id":"n01","type":"gene"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated genes for fasting glucose (EFO:0004465)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"EFO:0004465","id":"n00","type":"phenotypic_feature"},{"id":"n01","type":"gene"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated pathways for type 2 diabetes (EFO:0001360)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"EFO:0001360","id":"n00","type":"disease"},{"id":"n01","type":"pathway"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query
> 
> find associated pathways for fasting glucose (EFO:0004465)
> curl -X POST -H "Content-Type:application/json" -d '{"message":{"query_graph":{"edges":[{"id":"e00","source_id":"n00","target_id":"n01","type":"associated"}],"nodes":[{"curie":"EFO:0004465","id":"n00","type":"phenotypic_feature"},{"id":"n01","type":"pathway"}]}}}' https://translator.broadinstitute.org/genetics_data_provider/query

**Comment by amykglen** [comment]
does anyone have a suggestion for what EDAM data type we should use for the GeneticsKP's "Genetics-quantile" scores when loading them into an `EdgeAttribute`? (e.g., `EDAM:data_0951`)

example of such a score:
```
      {
        "id": "INTEGRATED_GENETICS_80254",
        "score": 0.96406529962174,
        "score_direction": "higher_is_better",
        "score_name": "Genetics-quantile",
        "source_id": "EFO:0001360",
        "target_id": "NCBIGene:85236",
        "type": "associated"
      }
```

**Comment by edeutsch** [comment]
I don't think EDAM has anything appropriate that I can find. I would suggest SIO:001414

https://www.ebi.ac.uk/ols/search?q=quantile

SIO is Michel Dumontier's product, so seems like an apt choice

**Comment by amykglen** [comment]
ok, I added @edeutsch's suggested data type (thanks!) and I think this integration can now be considered complete. some notes:

1. I opted to stick with the method of sequentially querying the GeneticsKP API in the case of multiple input curies, rather than attempting to use their current method for batch querying. seems it makes more sense to wait until there is a standard way of sending batch queries via TRAPI.
2. one quirk when using the GeneticsKP is that it requires qnode types and is sensitive to the use of `disease` vs. `phenotypic_feature`: that is, for example, a query that involves `add_qnode(name=type 2 diabetes mellitus, type=disease, id=n00)` _will_ return results, but `add_qnode(name=type 2 diabetes mellitus, type=phenotypic_feature, id=n00)` will not. and the reverse can be true depending on the concept you're querying for (`phenotypic_feature` will return results but `disease` won't). I could code in a workaround here that tries `phenotypic_feature` if `disease` didn't work and vice versa. though I'm not sure if that's something we should fix on our end, vs. perhaps requesting that they don't require types for "pinned" qnodes?
3. it appears some changes were made to their API (yesterday, I think?) and now there are more scores returned than only MAGMA p-value and genetics-quantile. currently these are grabbed if `include_all_scores=true`, but the `ARAXRanker` doesn't know what to do with these scores yet. (I'll create a separate issue on that topic.)

documentation has been updated and all pytests are passing for me in `master`.
